enable grafana_annotations callback sending annotations to multiple panels

emschu opened the pull request:
##### SUMMARY
Enables grafana_annotations callback plugin sending annotations to multiple panels of a dashboard.
+ small refactoring of dashboardId + panelId handling
+ plugin doc modifications

The user is able to set for "panelId" whether nothing, a single integer or a list of integers separated by comma.

##### ISSUE TYPE
- Feature Pull Request

##### COMPONENT NAME
grafana_annotations callback plugin

##### ADDITIONAL INFORMATION

Commit:
enable grafana_annotations callback plugin message sending to multiple panels

Commit:
remove new line at end of file

rrey commented on lib/ansible/plugins/callback/grafana_annotations.py:
```diff
@@ -249,15 +249,28 @@ def v2_runner_on_failed(self, result, **kwargs):
         if self.panel_id:
-            data["panelId"] = int(self.panel_id)
-        self._send_annotation(json.dumps(data))
+            uses_panels = True
```
I think a function handling the `self.panel_id` would help make the code clearer.

The function could always return a list of integer or an empty list.
Then you can iterate on the list to send the annotation. When the list is empty nothing would be done and the check line 268 is no more required.

emschu replied:
Thank you for your suggestion. Its done. Do you know how to fix this deprecation warning?

> [DEPRECATION WARNING]: grafana_annotations callback, does not support setting 'options', it will work for now,  but this will be required in the future and should be updated,  see the 2.4 porting guide for details.. This feature will be removed in version 2.9. Deprecation warnings can be disabled by setting  deprecation_warnings=False in ansible.cfg.

rrey replied:
no, I forgot this was around ...
I think I'll ask for tips on IRC

rrey replied:
Got it: https://github.com/ansible/ansible/pull/57629
My best patch ever :)

rrey reviewed (changes requested):
With that little change made it will be ok for me. Thanks for the contribution  !

Commit:
restructuring code of grafana_annotations cb plugin

emschu committed:
add missing self keyword

emschu commented:
ready_for_review

rrey reviewed (approved):
lgtm

rrey commented:
ship_it

emschu commented:
bot_status

mscherer commented:
shipit

rrey commented:
bot_status

pilou- commented on lib/ansible/plugins/callback/grafana_annotations.py:
```diff
@@ -93,20 +97,24 @@
         ini:
           - section: callback_grafana_annotations
             key: grafana_panel_id
+        type: string
```
would not that be better to use `type: list`

rrey replied:
Hi @pilou- 
I don’t know if the author will answer after so much time.
I could make the change in another PR but how would a list be provided through env vars ?

emschu replied:
@pilou- thx for your review.
@rrey thank you, I can do it too :) but at first I have the same question like you as there are no examples in callback plugin code.
No results for: `grep -ir "type: list" plugins/callback`                                                                                                  `

gundalow commented:
rebuild_merge

pilou- committed:
grafana callback: use list for panel_ids

pilou- committed:
grafana callback: convert panel_id to integer

annotations HTTP API states that panelId are integer (same as
dashboardId which is also converted).

pilou- commented:
@rrey @emschu [ensure_type method]() will convert a string (composed of one single element or composed of comma separated elements) into a list. 

1. [`ensure_type`](https://github.com/ansible/ansible/blob/d829a50a5b7030b82ec2d15a33540225209453d4/lib/ansible/config/manager.py#L54) is called by [`ConfigManager.get_config_value_and_origin`](https://github.com/ansible/ansible/blob/b20a0f4831da5abc5f1a03d60062538ddbaf73cb/lib/ansible/config/manager.py#L420)
2. which is called by [`ConfigManager.get_config_value`](https://github.com/ansible/ansible/blob/b20a0f4831da5abc5f1a03d60062538ddbaf73cb/lib/ansible/config/manager.py#L408)
3. which is called by [`ConfigManager.get_plugin_options`](https://github.com/ansible/ansible/blob/b20a0f4831da5abc5f1a03d60062538ddbaf73cb/lib/ansible/config/manager.py#L343)
4. which is called by [`CallbackBase.set_options`](https://github.com/ansible/ansible/blob/8d0c193b251051e459513614e6f7212ecf9f0922/lib/ansible/plugins/callback/__init__.py#L93). 

Sadly ` elements` (the type of elements for the list) can not be specified with plugins (because [`module_utils` utilities aren't reused with plugins](https://github.com/ansible/ansible/blob/d829a50a5b7030b82ec2d15a33540225209453d4/lib/ansible/config/manager.py#L55)).

Here is a proposal PR which uses `type: list` and converts `panelId` to integer: https://github.com/emschu/ansible/pull/1.

emschu commented:
@pilou- Thank you very much about this information and the code. I merged your proposal PR and marked your issue as resolved.
@rrey

rrey commented:
@emschu the CI failed because of timeouts, can you push again to re-trigger the CI ?

gundalow commented:
I've triggered a rerun of CI

rrey commented:
ship_it
